src/main.rs

```diff
@@ -403,6 +407,82 @@ async fn consume_messenger_events(
     Ok(())
 }
 
+#[async_trait]
+trait SendCustomMessage {
+    async fn send_custom_message(
+        &mut self,
+        request: SendCustomMessageRequest,
+    ) -> Result<SendCustomMessageResponse, Status>;
+}
+
+struct CustomMessenger<'a> {
+    client: &'a mut LightningClient,
+}
+
+#[async_trait]
+impl<'a> SendCustomMessage for CustomMessenger<'a> {
+    async fn send_custom_message(
+        &mut self,
+        request: SendCustomMessageRequest,
+    ) -> Result<SendCustomMessageResponse, Status> {
+        match self.client.send_custom_message(request).await {
+            Ok(resp) => Ok(resp.into_inner()),
+            Err(status) => Err(status),
+        }
+    }
+}
+
+// relay_messenger_events is responsible for polling the onionmessenger for new messages from peers. If a new OM
+// turns up, it will pass it along to lnd.
+async fn relay_messenger_events(
+    current_peers: &HashMap<PublicKey, bool>,
+    onion_messenger: &dyn OnionMessageHandler,
+    mut ln_client: impl SendCustomMessage,
+    mut exit: Receiver<()>,
+) {
+    let mut interval = time::interval(MSG_POLL_INTERVAL);
+
+    loop {
+        select! (
+            // Select biased so that we'll always check our exit signal before attempting to receive. This allows more
+            // deterministic tests, and ensures that the producer will exit when requested (and won't queue up a series
+            // of events that can't be consumed, possibly blocking if the channel buffer is small).
+            biased;
+
+            _ = exit.recv() => {
+                info!("Peer events received signal to quit");
+                return
+            }
+
+            _ = interval.tick() => {
+                for peer in current_peers.keys() {
+                    while let Some(msg) = onion_messenger.next_onion_message_for_peer(*peer) {
+                        debug!("received new om {msg:?} from peer {}", *peer);
+
+                        let mut buf = vec![];
+                        _ = msg.write(&mut buf);
+
+                        // Relay this message to LND.
+                        let req = tonic_lnd::lnrpc::SendCustomMessageRequest{
+                            peer: peer.serialize().to_vec(),
+                            r#type: ONION_MESSAGE_TYPE,
+                            data: buf,
+                        };
+
+                        // TODO: To improve resilience, retry this call in the event of a temporary connection error.
```

Comment:
I was thinking more in terms of -- if there's a transient issue with LND as a whole, like internet is wonky for a bit, and the client gets a connectionrefused error or unavailable error or something like that, we should retry the call

Reply:
Ah gotchu - makes sense! Could match on the grpc error codes LND returns or something of the sort 👍

Reply:
Cool cool, I'm thinking of leaving this to a follow-up PR because I think there are a couple of other spots we might want to repeat this retrying pattern -- like when calling `get_info` and `list_peers`

Reply:
Definitely a follow up, would be good to do a full sweep once off for all the various LND calls then enforce proper handling from there onwards.